packages/compose/src/hooks/use-debounce/index.js

```diff
@@ -9,17 +9,31 @@ import { useMemoOne } from 'use-memo-one';
  */
 import { useEffect } from '@wordpress/element';
 
+/* eslint-disable jsdoc/valid-types */
+/**
+ * @template T
+ * @typedef {T extends [any, ...infer R] ? R : never} TailParameters
+ */
+
+// We want to keep the type of the function that is passed in but the rest we don't
+// really care about, so this allows us to ignore the details of the rest of the paramters
+// while still maintaining the proper generic type for `debounce` where the returned
+// function matchest the signature of the passed in function.
+/** @typedef {TailParameters<Parameters<typeof debounce>>} DebounceTailParameters */
+
 /**
  * Debounces a function with Lodash's `debounce`. A new debounced function will
  * be returned and any scheduled calls cancelled if any of the arguments change,
  * including the function to debounce, so please wrap functions created on
  * render in components in `useCallback`.
  *
- * @param {...any} args Arguments passed to Lodash's `debounce`.
+ * @template {(...args: any[]) => void} TFunc
+ * @param {[TFunc, ...DebounceTailParameters]} args Arguments passed to Lodash's `debounce`.
  *
- * @return {Function} Debounced function.
+ * @return {TFunc & import('lodash').Cancelable} Debounced function.
  */
 export default function useDebounce( ...args ) {
+	/* eslint-enable jsdoc/valid-types */
```

Comment:
Should the `eslint` rule be outside of the `useDebounce`'s function body?

```suggestion
/* eslint-enable jsdoc/valid-types */
export default function useDebounce( ...args ) {
```

Reply:
No because that causes the comment applied to `useDebounce` to be the `eslint-enable` comment instead of its actual JSDoc comment.

Reply:
Makes sense, thank you for the explanation!